pkg/shares/shares_test.go

```diff
@@ -98,3 +98,181 @@ func generateRandomBlobOfShareCount(count int) coretypes.Blob {
 func rawBlobSize(totalSize int) int {
 	return totalSize - DelimLen(uint64(totalSize))
 }
+
+func TestSequenceLen(t *testing.T) {
+	type testCase struct {
+		name         string
+		share        Share
+		wantLen      uint64
+		wantNumBytes int
+		wantErr      bool
+	}
+	firstShare := []byte{
+		1, 1, 1, 1, 1, 1, 1, 1, // namespace
+		1,           // info byte
+		10, 0, 0, 0, // sequence len
+		1, 2, 3, 4, 5, 6, 7, 8, 9, 10, // data
+	}
+	firstShareWithLongSequence := []byte{
+		1, 1, 1, 1, 1, 1, 1, 1, // namespace
+		1,            // info byte
+		195, 2, 0, 0, // sequence len
+	}
+	continuationShare := []byte{
+		1, 1, 1, 1, 1, 1, 1, 1, // namespace
+		0,  // info byte
+		10, // sequence len
+	}
+	compactShare := []byte{
+		0, 0, 0, 0, 0, 0, 0, 1, // namespace
+		1,           // info byte
+		10, 0, 0, 0, // sequence len
+	}
+	noInfoByte := []byte{
+		0, 0, 0, 0, 0, 0, 0, 1, // namespace
+	}
+	noSequenceLen := []byte{
+		0, 0, 0, 0, 0, 0, 0, 1, // namespace
+		1, // info byte
+	}
+	testCases := []testCase{
+		{
+			name:         "first share",
+			share:        firstShare,
+			wantLen:      10,
+			wantNumBytes: 1,
+			wantErr:      false,
+		},
+		{
+			name:         "first share with long sequence",
+			share:        firstShareWithLongSequence,
+			wantLen:      323,
+			wantNumBytes: 2,
+			wantErr:      false,
+		},
+		{
+			name:         "continuation share",
+			share:        continuationShare,
+			wantLen:      0,
+			wantNumBytes: 0,
+			wantErr:      false,
+		},
+		{
+			name:         "compact share",
+			share:        compactShare,
+			wantLen:      10,
+			wantNumBytes: 4,
+			wantErr:      false,
+		},
+		{
+			name:         "no info byte returns error",
+			share:        noInfoByte,
+			wantLen:      0,
+			wantNumBytes: 0,
+			wantErr:      true,
+		},
+		{
+			name:         "no sequence len returns error",
+			share:        noSequenceLen,
+			wantLen:      0,
+			wantNumBytes: 0,
+			wantErr:      true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			len, numBytes, err := tc.share.SequenceLen()
+
+			if tc.wantErr {
+				assert.Error(t, err)
+				return
+			}
+			assert.Equal(t, tc.wantLen, len)
+			assert.Equal(t, tc.wantNumBytes, numBytes)
+		})
+	}
+}
+
+func TestRawData(t *testing.T) {
+	type testCase struct {
+		name    string
+		share   Share
+		want    []byte
+		wantErr bool
+	}
+	firstSparseShare := []byte{
+		1, 1, 1, 1, 1, 1, 1, 1, // namespace
+		1,                             // info byte
+		10,                            // sequence len
+		1, 2, 3, 4, 5, 6, 7, 8, 9, 10, // data
+	}
+	continuationSparseShare := []byte{
+		1, 1, 1, 1, 1, 1, 1, 1, // namespace
+		0,                             // info byte
+		1, 2, 3, 4, 5, 6, 7, 8, 9, 10, // data
+	}
+	firstCompactShare := []byte{
+		0, 0, 0, 0, 0, 0, 0, 1, // namespace
+		1,           // info byte
+		10, 0, 0, 0, // sequence len
+		1, 2, 3, 4, 5, 6, 7, 8, 9, 10, // data
+	}
```

Comment:
More of a side question: aren't there supposed to be two reserved bytes as well?

Reply:
Great question and yes! This unit test still passes because the two reserved bytes are treated as part of the data (i.e. `1, 2` are the reserved bytes in this test) but this may make it clearer:

```golang
	firstCompactShare := []byte{
		0, 0, 0, 0, 0, 0, 0, 1, // namespace
		1,           // info byte
		10, 0, 0, 0, // sequence len
                15, 0, // reserved bytes
		1, 2, 3, 4, 5, 6, 7, 8, 9, 10, // data
	}
```

Reply:
Explicitly added in https://github.com/celestiaorg/celestia-app/pull/1096/commits/a2b374d04cf52e412c8d77618b033bf1ccf682f4

FWIW I don't have signal on whether consumers of `share.RawData()` want the reserved bytes included or not. I elected to include them now and we can revisit if we get signal otherwise.

Reply:
Who are the consumers of `share.RawData()`? I assume they'll want all the information they need to parse the txs within the share so perhaps everything minus the namespace itself